builder/builder.go

```diff
@@ -443,3 +454,92 @@ func parseImageNameFromArgs(cmdArgs string) string {
 	}
 	return cmdArgs[:idx]
 }
+
+// prepareVolumeSource creates and populates the host file and volume for the specified source type
+func (b *Builder) prepareVolumeSource(ctx context.Context, volMount *volume.Volume) error {
+	switch {
+	case volMount.Source.Secret != nil:
+		if err := b.createSecretFiles(ctx, volMount); err != nil {
+			return err
+		}
+		return b.populateSecretVolume(ctx, volMount)
+	default:
+		return errors.New("source type not supported yet")
+	}
+}
+
+// createSecretFiles creates necessary files for source type Secret
+func (b *Builder) createSecretFiles(ctx context.Context, volMount *volume.Volume) error {
+	var args []string
+	var sb strings.Builder
+	if runtime.GOOS == util.WindowsOS {
+		args = []string{"powershell.exe", "-Command"}
+		sb.WriteString("mkdir " + volMount.Name)
+		log.Println("making directory " + volMount.Name)
+	} else {
+		args = []string{"/bin/sh", "-c"}
+		sb.WriteString("mkdir " + volMount.Name + " && ")
+	}
+	for _, values := range volMount.Source.Secret {
+		for k, v := range values {
+			val := v
+			decoded, err := base64.StdEncoding.DecodeString(val)
+			if err != nil {
+				return errors.New("failed to decode Base64 value. please make sure value provided is Base64 encoded")
+			}
+			val = string(decoded)
+			if runtime.GOOS == util.WindowsOS {
+				sb.WriteString("; Add-Content -Path ")
+				sb.WriteString(volMount.Name + "/" + k)
+				sb.WriteString(" -Value @\"\n")
+				sb.WriteString(val)
+				sb.WriteString("\n\"@")
+			} else {
+				sb.WriteString("cat >> ")
+				sb.WriteString(volMount.Name + "/" + k)
+				sb.WriteString(" <<EOL\n")
+				sb.WriteString(val)
+				sb.WriteString("\nEOL\n")
+			}
+		}
+	}
+	args = append(args, sb.String())
+	var buf bytes.Buffer
+	if err := b.procManager.Run(ctx, args, nil, &buf, &buf, ""); err != nil {
+		return errors.Wrapf(err, "failed to write value, %s", buf.String())
+	}
+	log.Println("buffer: ", buf.String())
+	log.Println("Created new file(s) for volume: ", volMount.Name)
+	return nil
+}
+
+// populateSecretVolume mounts all files of type source Secret generated into a volume
+func (b *Builder) populateSecretVolume(ctx context.Context, volMount *volume.Volume) error {
+	var dataContainerArgs []string
+	var dataSB strings.Builder
+	if runtime.GOOS == util.WindowsOS {
+		dataContainerArgs = []string{"powershell.exe", "-Command"}
+		dataSB.WriteString("docker run --rm -v " + b.workspaceDir + ":c:\\source -v ")
+		dataSB.WriteString(volMount.Name + ":c:\\dest -w c:\\source ")
+		dataSB.WriteString(nanoServerImageName + " cmd.exe /c copy c:\\source\\" + volMount.Name + " c:\\dest")
+	} else {
+		dataContainerArgs = []string{"/bin/sh", "-c"}
+		dataSB.WriteString("docker run --rm -v " + b.workspaceDir + ":/source -v ")
+		dataSB.WriteString(volMount.Name + ":/dest -w /source alpine cp ")
```

Comment:
>alpine [](start = 56, length = 6)

is it possible to use bash? 

[https://github.com/Azure/acr-builder/blob/master/builder/constants_linux.go](https://github.com/Azure/acr-builder/blob/master/builder/constants_linux.go)